# Clustering Toronto Neighborhoods 


In this notebook, we will be scraping a Wikipedia page for the Postal Codes of Canada. Unfortunately, the dataset is not easily accessible so we will have to wrangle and process the data.

First, we will import requests so we can acquire the link that holds the data we are intersted in.

After importing requests and assigning the Wiki link, we will need to **get** the **raw_Toronto** into **text** format

In [1]:
import requests
Toronto_link ="https://en.m.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

In [2]:
raw_Toronto=requests.get(Toronto_link).text

Now, we must import **Beautiful Soup** in order to help parse html from Wikipedia.
The **lxml** parser is used to create the soup as we scrape the webpage. Powerful 
Beautiful Soup allows us to immediately find the **table** using the **find** feature.

In [3]:
from bs4 import BeautifulSoup
soup = BeautifulSoup(raw_Toronto, 'lxml')
table = soup.find('table')

**Prettify** is then used since the data would be a bit raw and hard to read.  I used it in order to identify the proper tags for extraction.

In [4]:
TorTable=table.prettify()

Now, we import **pandas** in order to be able to read the Prettified ’TorTable’ as html into dfs.  **Header** is selected in order to identify the first row as the header row *Postcode, Borough, Neighbourhood*. The **dfs** DataFrame is placed into a single object which I will assign to **TorTablePost** for filtering.  Next, we filter TorTablePost removing any rows that do not have a Borough assigned a value.

In [5]:
import pandas as pd
dfs = pd.read_html(TorTable, header=0)
TorTablePost = dfs[0]
TorT=TorTablePost[TorTablePost.Borough != 'Not assigned']

Now that we do not have any missing values in the *Postcode* and *Borough* columns, the next thing we will do is to change the **Not Assigned** values in **Neighbourhood** to the name of the **Borough**. In order to accomplish this, **Numpy** is imported to adjust the **Not Assigned** values to *np.nan*. Then, we use the **fillna** method to simultaneously filter the **Neighbourhood** column and add the values from **Borough** into the *NaN* slots.

In [6]:
import numpy as np
TorT.replace('Not assigned', np.nan, inplace=True)
TorT['Neighbourhood'].fillna(TorT['Borough'], inplace=True)

/opt/conda/envs/DSX-Python35/lib/python3.5/site-packages/ipykernel/__main__.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from ipykernel import kernelapp as app
/opt/conda/envs/DSX-Python35/lib/python3.5/site-packages/pandas/core/generic.py:4355: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)


One of the last adjustments sought after is to compile all the distinct **Neighborhoods** with the same *PostCode* into its like row. In order to do so, we will group *Postcode* and *Borough* together, and find each individual value of **Neighborhood** and **join** them together with a comma. Finally, reset the index for the new dataframe.

In [7]:
TorGroup=TorT.groupby(['Postcode','Borough'])['Neighbourhood'].apply(', '.join).reset_index()

As we want the column headers to read *PostalCode, Borough, and Neighborhood*, we reassign the columns to these names.

In [8]:
TorGroup.columns = ['PostalCode', 'Borough', 'Neighborhood']

In [9]:
TorGroup.shape

(103, 3)

The Dataframe consists of **103** rows and **3** columns

The Latitude and Longitude data for Toronto is stored at the link below, so that will be loaded to be merged with our data

In [10]:
TorLLink = "https://cocl.us/Geospatial_data"
df = pd.read_csv(TorLLink, header=0)

The index is reset and set to Postal Code

In [11]:
TorLL=TorGroup.set_index('PostalCode').join(df.set_index('Postal Code')).reset_index()
TorLL.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

We will only explore the boroughs in Toronto, so adding a filter to get that Toronto data only.

In [12]:
toronto_data = TorLL[TorLL['Borough'].str.contains("Toronto")].reset_index(drop=True)
toronto_data.head()

PostalCode          Borough                    Neighborhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [13]:
print('The dataframe has {} boroughs across {} postal codes.'.format(
     len(toronto_data['Borough'].unique()),
     toronto_data.shape[0]))

The dataframe has 4 boroughs across 38 postal codes.


So we have 4 boroughs—East, Central, West & Downtown Toronto— and 38 postal codes amonst these boroughs

The next set of analysis requires exploring maps so we will import the dependencies.

In [14]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    geographiclib: 1.49-py_0   conda-forge
    geopy:         1.19.0-py_0 conda-forge

geographiclib- 100% |################################| Time: 0:00:00  24.07 MB/s
geopy-1.19.0-p 100% |################################| Time: 0:00:00  35.89 MB/s
Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    altair:  2.2.2-py35_1 conda-forge
    branca:  0.3.1-py_0   conda-forge
    folium:  0.5.0-py_0   conda-forge
    vincent: 0.4.4-py_1   conda-forge

altair-2.2.2-p 100% |################################| Time: 0:00:00  54.68 MB/s
branca-0.3.1-p 100% |################################| Time: 0:00:00  35.07 MB/s
vincent-0.4.4- 100% |###################

In order to correctly locate and center on Toronto, we set the Address to Toronto, use the geolocator and retrieve the geographical coordinates

In [15]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


Now it is time to make the first visual for the map, marking the map with the neighborhood labels.

In [16]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough'], toronto_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

We store our Foursquare credentials to help us create the url later.

In [17]:
CLIENT_ID = 'F4XRVGG0NQPWPBGMBL5SULBTIQB4X4IXIEL5XZE4TNW4WCJE' # your Foursquare ID
CLIENT_SECRET = '5ETJCTC2N15PIRCCFYMYA0CWXJXISBWLKQEXK0IGABAEKGVH' # your Foursquare Secret
VERSION = '20190325' # Foursquare API version

print('Your credentials:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentials:
CLIENT_ID: F4XRVGG0NQPWPBGMBL5SULBTIQB4X4IXIEL5XZE4TNW4WCJE
CLIENT_SECRET:5ETJCTC2N15PIRCCFYMYA0CWXJXISBWLKQEXK0IGABAEKGVH


A function is created in order to get the nearby venues for each neighborhood

In [18]:
LIMIT = 100
radius = 500


def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

A new dataframe *toronto_venues* is created to store the data created in the previous step.

In [19]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
The Beaches West, India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North, Forest Hill West
The Annex, North Midtown, Yorkville
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
Dovercourt Village, Dufferin
Little Portugal, Trinity
Brockton, Exhibition Place, Parkdale Village
High Park, The 

The previous url I made was local within the neighborhood function, this one will provide the results for the rest of the rest of the data

In [20]:
# type your answer here
LIMIT = 100 
radius = 500

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID,
    CLIENT_SECRET,
    VERSION,
    lat,
    lng,
    radius,
    LIMIT)

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5c9f6bbc9fb6b73bddc6e0f7'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-58d4650e2f91cb5ccec03447-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/brewery_',
          'suffix': '.png'},
         'id': '50327c8591d4c4b30a586d5d',
         'name': 'Brewery',
         'pluralName': 'Breweries',
         'primary': True,
         'shortName': 'Brewery'}],
       'id': '58d4650e2f91cb5ccec03447',
       'location': {'address': '1001 Eastern Ave',
        'cc': 'CA',
        'city': 'Toronto',
        'country': 'Canada',
        'crossStreet': 'Woodfield Rd',
        'distance': 162,
        'formattedAddress': ['1001 Eastern Ave (Woodfield Rd)',
         'Toronto ON M4L 1A8',
         'Canada'],
        'labeledLatLngs': [{'label'

The category of each venue is extracted from the data.

In [21]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we clean the json and create a new dataframe

In [22]:
import json
from pandas.io.json import json_normalize

venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name      categories        lat        lng
0       Rorschach Brewing Co.         Brewery  43.663483 -79.319824
1  Leslieville Farmers Market  Farmers Market  43.664901 -79.319784
2                The Sidekick      Comic Shop  43.664484 -79.325162
3      Queen Margherita Pizza     Pizza Place  43.664685 -79.324164
4                 Chino Locos   Burrito Place  43.664653 -79.325584

A count function is created in order to identify how many venues were returned for each neighborhood

In [23]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            55   
Brockton, Exhibition Place, Parkdale Village                           22   
Business Reply Mail Processing Centre 969 Eastern                      17   
CN Tower, Bathurst Quay, Island airport, Harbou...                     14   
Cabbagetown, St. James Town                                            46   
Central Bay Street                                                     83   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               16   
Church and Wellesley                                                   86   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             37   
Davisville North                                                        8   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     14   
Design Exchange, Toronto Dominion Centre                              100   
Dovercourt Village, Dufferin                                           18   
First Canadian Place, Underground city                                100   
Forest Hill North, Forest Hill West                                     4   
Harbord, University of Toronto                                         35   
Harbourfront East, Toronto Islands, Union Station                     100   
Harbourfront, Regent Park                                              46   
High Park, The Junction South                                          24   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               65   
Moore Park, Summerhill East                                             2   
North Toronto West                                                     20   
Parkdale, Roncesvalles                                                 16   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede, Swansea                                                     37   
Ryerson, Garden District                                              100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        94   
Studio District                                                        39   
The Annex, North Midtown, Yorkville                                    24   
The Beaches                                                             5   
The Beaches West, India Bazaar                                         21   
The Danforth West, Riverdale                                           42   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Berczy Park                                                             55   
Brockton, Exhibition Place, Parkdale Village                            22   
Business Reply Mail Processing Centre 969 Eastern                       17   
CN Tower, Bathurst Quay, Island airport, Harbou...                      14   
Cabbagetown, St. James Town                                             46   
Central Bay Street                                                      83   
Chinatown, Grange Park, Kensington Market                              100   
Christie                                                                16   
Church and Wellesley                                                    86   
Commerce Court, Victoria Hotel     

237 categories are found

In [24]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 237 uniques categories.


The *toronto_onehot* dataframe is created in order to create a line for each neighboorhood with a venue from a pertaining category

In [25]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0            0                  0               0                  0        0   
1            0                  0               0                  0        0   
2            0                  0               0                  0        0   
3            0                  0               0                  0        0   
4            0                  0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Beer Bar  Beer Store  \
0                 0                   0      0         0           0   
1                 0                   0      0         0           0   
2                 0                   0      0         0           0   
3                 0                   0      0         0           0   
4                 0                   0      0         0           0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0       0              0          0         0   
1                   0       0              0          0         0   
2                   0       0              0          0         0   
3                   0       0              0          0         0   
4                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0        0     0   

   Cajun / Creol

The results are then grouped together by neighborhood

In [26]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                            Adelaide, King, Richmond     0.000000   
1                                         Berczy Park     0.000000   
2        Brockton, Exhibition Place, Parkdale Village     0.045455   
3   Business Reply Mail Processing Centre 969 Eastern     0.058824   
4   CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000   
5                         Cabbagetown, St. James Town     0.000000   
6                                  Central Bay Street     0.012048   
7           Chinatown, Grange Park, Kensington Market     0.000000   
8                                            Christie     0.000000   
9                                Church and Wellesley     0.023256   
10                     Commerce Court, Victoria Hotel     0.000000   
11                                         Davisville     0.000000   
12                                   Davisville North     0.000000   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...     0.000000   
14           Design Exchange, Toronto Dominion Centre     0.000000   
15                       Dovercourt Village, Dufferin     0.000000   
16             First Canadian Place, Underground city     0.000000   
17                Forest Hill North, Forest Hill West     0.000000   
18                     Harbord, University of Toronto     0.028571   
19  Harbourfront East, Toronto Islands, Union Station     0.000000   
20                          Harbourfront, Regent Park     0.000000   
21                      High Park, The Junction South     0.000000   
22                                      Lawrence Park     0.000000   
23                           Little Portugal, Trinity     0.015385   
24                        Moore Park, Summerhill East     0.000000   
25                                 North Toronto West     0.050000   
26                             Parkdale, Roncesvalles     0.000000   
27                                           Rosedale     0.000000   
28                                           Roselawn     0.000000   
29                                 Runnymede, Swansea     0.000000   
30                           Ryerson, Garden District     0.000000   
31                                     St. James Town     0.000000   
32                    Stn A PO Boxes 25 The Esplanade     0.000000   
33                                    Studio District     0.025641   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                     The Beaches West, India Bazaar     0.000000   
37                       The Danforth West, Riverdale     0.023810   

    Accessories Store  Adult Boutique  Afghan Restaurant   Airport  \
0                0.01        0.000000           0.000000  0.000000   
1                0.00        0.000000           0.000000  0.000000   
2                0.00        0.000000           0.000000  0.000000   
3                0.00        0.000000           0.000000  0.000000   
4                0.00        0.000000           0.000000  0.071429   
5                0.00        0.000000           0.000000  0.000000   
6                0.00        0.000000           0.000000  0.000000   
7                0.00        0.000000           0.000000  0.000000   
8                0.00        0.000000           0.000000  0.000000   
9                0.00        0.011628           0.011628  0.000000   
10               0.00        0.000000           0.000000  0.000000   
11               0.00        0.000000           0.000000  0.000000   
12               0.00        0.000000           0.000000  0.000000   
13               0.00        0.000000           0.000000  0.000000   
14               0.00        0.000000           0.000000  0.000000   
15               0.00        0.000000           0.000000  0.000000   
16               0.00        0.000000           0.000000  0.000000   
17               0.00        0.000000           0.

The top 5 venues are recovered according to the frequency

In [27]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
             venue  freq
0      Coffee Shop  0.06
1              Bar  0.04
2       Steakhouse  0.04
3             Café  0.04
4  Thai Restaurant  0.04


----Berczy Park----
            venue  freq
0     Coffee Shop  0.07
1    Cocktail Bar  0.05
2      Restaurant  0.04
3  Farmers Market  0.04
4            Café  0.04


----Brockton, Exhibition Place, Parkdale Village----
                venue  freq
0      Breakfast Spot  0.09
1         Coffee Shop  0.09
2                Café  0.09
3       Burrito Place  0.05
4  Falafel Restaurant  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0         Yoga Studio  0.06
1       Auto Workshop  0.06
2       Garden Center  0.06
3              Garden  0.06
4  Light Rail Station  0.06


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
              venue  freq
0  Airport Terminal  0.14
1   Airport Service  0.14
2    

The data is sorted in descending order 

In [28]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

And we create a function to store the top 10 venues for each neighborhood into a dataframe

In [29]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village        Breakfast Spot   
3  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
4  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Bar            Steakhouse                  Café   
1          Cocktail Bar            Restaurant           Cheese Shop   
2                  Café           Coffee Shop           Yoga Studio   
3         Auto Workshop            Smoke Shop               Brewery   
4      Airport Terminal       Airport Service                 Plane   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0       Thai Restaurant                 Hotel      Asian Restaurant   
1                Bakery        Farmers Market                   Pub   
2               Stadium         Burrito Place            Restaurant   
3    Light Rail Station        Farmers Market                   Spa   
4         Boat or Ferry      Sculpture Garden       Harbor / Marina   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0      Sushi Restaurant                Bakery    American Restaurant  
1                  Café    Seafood Restaurant             Steakhouse  
2  Caribbean Restaurant          Climbing Gym                    Gym  
3  Fast Food Restaurant         Burrito Place             Restaurant  
4              Boutique          Airport Gate                Airport

Now is the time to decide how many clusters to be used to properly examine the data.  

In [42]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

The clustering labels are added to the dataframe, and the data from the orignal dataframe is merged with the venues data

(I have added a **drop** column function to allow me to consider other cluster sizes without having to re-run the entire data set)

In [43]:
# add clustering labels, drop is added for reassessing the number of clusters.
neighborhoods_venues_sorted.drop(['Cluster Labels'], axis=1, inplace=True)
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode          Borough                    Neighborhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               4           Coffee Shop     Health Food Store   
1 -79.352188               0      Greek Restaurant           Coffee Shop   
2 -79.315572               0                  Park        Sandwich Place   
3 -79.340923               0                  Café           Coffee Shop   
4 -79.388790               2  Gym / Fitness Center              Bus Line   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Pub         Women's Store    Dim Sum Restaurant   
1        Ice Cream Shop    Italian Restaurant             Bookstore   
2             Pet Store    Italian Restaurant     Food & Drink Shop   
3                Bakery    Italian Restaurant             Gastropub   
4                  Park           Swim School    Falafel Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0    Falafel Restaurant           Event Space  Ethiopian Restaurant   
1           Yoga Studio                Bakery            Sports Bar   
2      Sushi Restaurant                   Pub        Ice Cream Shop   
3   American Restaurant           Yoga Studio           Fish Market   
4           Event Space  Ethiopian Restaurant     Electronics Store   

         9th Most Common Venue       10th Most Common Venue  
0            Electronics Store  Eastern European Restaurant  
1                          Spa                    Juice Bar  
2            Fish & Chips Shop                Burrito Place  
3                    Juice Bar              Coworking Space  
4  Eastern European Restaurant          Dumpling Restaurant

Next, the map is created showing the clusters. 5 clusters shows a fair enough representation of the data, clustering one large group together.  In order to break up that cluster, we would have to increase the amount of clusters to 10 and that could constitute an overfit in the data.

In [44]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Cluster 0 consists primarily of local hangouts (coffe shops & cafes)

In [33]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1       East Toronto               0      Greek Restaurant   
2       East Toronto               0                  Park   
3       East Toronto               0                  Café   
5    Central Toronto               0                   Gym   
6    Central Toronto               0           Coffee Shop   
7    Central Toronto               0        Sandwich Place   
11  Downtown Toronto               0            Restaurant   
12  Downtown Toronto               0           Coffee Shop   
13  Downtown Toronto               0           Coffee Shop   
14  Downtown Toronto               0           Coffee Shop   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0           Coffee Shop   
17  Downtown Toronto               0           Coffee Shop   
18  Downtown Toronto               0           Coffee Shop   
19  Downtown Toronto               0           Coffee Shop   
20  Downtown Toronto               0           Coffee Shop   
21  Downtown Toronto               0           Coffee Shop   
23   Central Toronto               0                 Trail   
24   Central Toronto               0        Sandwich Place   
25  Downtown Toronto               0                  Café   
26  Downtown Toronto               0                  Café   
27  Downtown Toronto               0        Airport Lounge   
28  Downtown Toronto               0           Coffee Shop   
29  Downtown Toronto               0           Coffee Shop   
30  Downtown Toronto               0         Grocery Store   
31      West Toronto               0              Pharmacy   
32      West Toronto               0                   Bar   
33      West Toronto               0        Breakfast Spot   
34      West Toronto               0                   Bar   
35      West Toronto               0        Breakfast Spot   
36      West Toronto               0                  Café   
37      East Toronto               0           Yoga Studio   

   2nd Most Common Venue          3rd Most Common Venue  \
1            Coffee Shop                 Ice Cream Shop   
2         Sandwich Place                      Pet Store   
3            Coffee Shop                         Bakery   
5      Food & Drink Shop                           Park   
6    Sporting Goods Shop                 Clothing Store   
7           Dessert Shop                    Pizza Place   
11           Coffee Shop                           Park   
12   Japanese Restaurant                        Gay Bar   
13                  Café                         Bakery   
14        Clothing Store                           Café   
15            Restaurant                          Hotel   
16          Cocktail Bar                     Restaurant   
17        Sandwich Place             Italian Restaurant   
18                   Bar                     Steakhouse   
19              Aquarium                          Hotel   
20                  Café                          Hotel   
21                 Hotel                           Café   
23         Jewelry Store             Mexican Restaurant   
24                  Café                    Coffee Shop   
25                   Bar            Japanese Restaurant   
26                   Bar  Vegetarian / Vegan Restaurant   
27      Airport Terminal                Airport Service   
28            Restaurant                           Café   
29                  Café                          Hotel   
30                  Café                           Park   
31           Supermarket                 Discount Store   
32           Coffee Shop               Asian Restaurant   
33                  Café                    Coffee Shop   
34                  Café             Mexican Restaurant   
35             Gift Shop                   Burger Joint   
36           Coffee Shop                    Pizza Place   
37         Auto Workshop                     Smoke Shop   

        4th Most C

## Cluster 1 consists mainly of Eastern restaurants

In [34]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22  Central Toronto               1                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22   Filipino Restaurant        Farmers Market    Falafel Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
22           Event Space  Ethiopian Restaurant     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22  Eastern European Restaurant   Dumpling Restaurant             Donut Shop

## Cluster 2 consists of fitness and outdoors activities

In [35]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
4    Central Toronto               2  Gym / Fitness Center   
10  Downtown Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4               Bus Line                  Park           Swim School   
10            Playground                 Trail         Women's Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4     Falafel Restaurant           Event Space  Ethiopian Restaurant   
10          Dessert Shop           Event Space  Ethiopian Restaurant   

   8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
4      Electronics Store  Eastern European Restaurant    Dumpling Restaurant  
10     Electronics Store  Eastern European Restaurant    Dumpling Restaurant

## Cluster 3 is mainly playgrounds and social spaces

In [36]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
8  Central Toronto               3            Playground   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
8                   Gym          Dessert Shop    Falafel Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
8           Event Space  Ethiopian Restaurant     Electronics Store   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
8  Eastern European Restaurant   Dumpling Restaurant             Donut Shop

## Cluster 4 consists of Coffee Shops and pubs

In [37]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0     East Toronto               4           Coffee Shop   
9  Central Toronto               4           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0     Health Food Store                   Pub         Women's Store   
9                   Pub           Supermarket            Sports Bar   

  5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
0    Dim Sum Restaurant     Falafel Restaurant           Event Space   
9    Light Rail Station  Vietnamese Restaurant      Sushi Restaurant   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant  
9           Pizza Place     Convenience Store          Fried Chicken Joint